python/lsst/daf/butler/core/progress.py

```diff
@@ -0,0 +1,396 @@
+
+        Parameters
+        ----------
+        n : `int`, optional
+            Increment the progress bar by this many steps (defaults to ``1``).
+            Note that this is a relative increment, not an absolute progress
+            value.
+        """
+        pass
+
+
+class Progress:
+    """Public interface for reporting incremental progress in the butler and
+    related tools.
+
+    This class automatically creates progress bars (or not) depending on
+    whether a handle (see `ProgressHandler`) has been installed and the given
+    name and level are enabled.  When progress reporting is not enabled, it
+    returns dummy objects that can be used just like progress bars by calling
+    code.
+
+    Parameters
+    ----------
+    name : `str`
+        Name of the process whose progress is being reported.  This is in
+        general the name of a group of progress bars, not necessarily as single
+        one, and it should have the same form as a logger names: the
+        fully-qualified name of the class, method, or function, minus the
+        initial "lsst" (and without any "implemntation detail" packages or
+        modules).
+    level : `int`, optional
+        A `logging` level value (defaults to `logging.INFO`).  Progress
+        reporting is enabled if a logger with ``name`` is enabled for this
+        level, and a `ProgressHandler` has been installed.
+    """
+
+    def __init__(self, name: str, level: int = logging.INFO) -> None:
+        self._name = name
+        self._level = level
+
+    # The active handler is held in a ContextVar to isolate unit tests run
+    # by pytest-xdist.  If butler codes is ever used in a real multithreaded
+    # or asyncio application _and_ we want progress bars, we'll have to set
+    # up per-thread handlers or similar.
+    _active_handler: ClassVar[Optional[ProgressHandler]] = None
+
+    @classmethod
+    def set_handler(cls, handler: Optional[ProgressHandler]) -> None:
+        """Set the (global) progress handler to the given instance.
+
+        This should only be called in very high-level code that can be
+        reasonably confident that it will dominate its current process, e.g.
+        at the initialization of a command-line script or Jupyter notebook.
+
+        Parameters
+        ----------
+        handler : `ProgressHandler` or `None`
+            Object that will handle all progress reporting.  May be set to
+            `None` to disable progress reporting.
+        """
+        cls._active_handler = handler
+
+    def is_enabled(self) -> bool:
+        """Check whether this process should report progress.
+
+        Returns
+        -------
+        enabled : `bool`
+            `True` if there is a `ProgressHandler` set and a logger with the
+            same name and level as ``self`` is enabled.
+        """
+        if self._active_handler is not None:
+            logger = logging.getLogger(self._name)
```

Comment:
Does this work for `obs_base` which uses `lsst.log` and not `logging`?

Reply:
I'm not sure if it works to set the level for a "logger" that really corresponds to a progress bar via `lsst.log`; I used `logging` to do that (while still using `lsst.log` for _actual_ logs).  I think that's okay, because I'm really just using the logging system as a way to get a global mapping of name->level, but it probably is worth documenting.